clients/roscpp/src/libros/service_server_link.cpp

```diff
@@ -340,15 +340,15 @@ bool ServiceServerLink::call(const SerializedMessage& req, SerializedMessage& re
 
   bool immediate = false;
   {
-    boost::mutex::scoped_lock lock(call_queue_mutex_);
-
     if (connection_->isDropped())
     {
       ROSCPP_LOG_DEBUG("ServiceServerLink::call called on dropped connection for service [%s]", service_name_.c_str());
       info->call_finished_ = true;
       return false;
     }
 
+    boost::mutex::scoped_lock lock(call_queue_mutex_);
```

Comment:
How is this change related to the title of the PR?

Reply:
> How is this change related to the title of the PR?

After using lock before "drop_signal_(shared_from_this(), reason);", there will be a deadlock in two threads

```
thread1:
    ros::ServiceServerLink::call
        boost::mutex::scoped_lock lock(call_queue_mutex_);            # lock A

        ros::Connection::isDropped

            boost::recursive_mutex::scoped_lock lock(drop_mutex_);    # lock B
       
thread2:

    ros::Connection::drop
        boost::recursive_mutex::scoped_lock lock(drop_mutex_);        # lock B
       
        ros::ServiceServerLink::onConnectionDropped
            boost::mutex::scoped_lock lock(call_queue_mutex_);        # lock A
```